common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -409,5 +410,43 @@ trait MultipleReadersSingleWriterCache[W, Winfo] {
     }
 
     /** This is the backing store. */
-    private val cache: Cache[Entry] = LruCache(timeToLive = 5.minutes)
+    private val cache: Cache[Entry] = new CaffeineBackedCache(
+        Caffeine.newBuilder()
+            .expireAfterWrite(5, TimeUnit.MINUTES)
+            .softValues()
+            .build().asInstanceOf[CaffeineCache[Any, Future[Entry]]])
+}
+
+/**
+ * A **non-thread-safe** implementation of [[spray.caching.cache]] backed by
+ * [[com.github.benmanes.caffeine.cache.Caffeine]]. Locking access contract
+ * is provided by the MultipleReadersSingleWriterCache, hence this class is
+ * only to be used there.
+ *
+ * The implementation is mainly copied from [[spray.caching.LruCache]]
+ * and reimplemented using the slightly differing API of Caffeine.
+ *
+ * The unimplemented methods are not needed for our internal implementation.
+ */
+private class CaffeineBackedCache[V](store: CaffeineCache[Any, Future[V]]) extends Cache[V] {
+
+    def get(key: Any) = Option(store.getIfPresent(key))
+
+    def apply(key: Any, genValue: () ⇒ Future[V])(implicit ec: ExecutionContext): Future[V] = get(key).getOrElse {
```

Comment:
Yep, that's actually the one thing I thought our augmentation with SingleWriterMultipleReaders would help us.

Apparently there's an equivalent to `putIfAbsent`, updating.